### PR TITLE
Add BDD tests for Token Feature to RSTUF API

### LINKED ISSUE #281: Task: Add the Integration Tests/BDD for Token in the API
### What is the task about?

Currently, we have a Functional/BDD test in the umbrella repository (RSTUF) that [tests the Tokens as a Feature](https://github.com/vmware/repository-service-tuf/tree/main/tests/features/tokens).

It is currently a feature but not a cross-component feature.
The priority of our Functional/BDD Tests in the umbrella repository is to guarantee we don't break cross-functionality between the components.

This task is about moving this specific test inside the RSTUF API.
Instead of calling it Functional Test, we can call it Integration Test, as we will test the Authentication/Authorization integrated into the API.

All the external parts can be mocked, for example, submitting tasks to the broker.

### Parent feature

- https://github.com/vmware/repository-service-tuf/issues/41

### References

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by kairoaraujo**
The current BDD Tokens tests are part of integration tests in the RSTUF umbrella repository.
This BDD feature test doesn't belong to the integration tests once API Token authentication/authorization (auth) feature is a single component.

The BDD test is still valid once it describes very well the expected behavior.

This commit adds the test to the API and is part of the https://github.com/repository-service-tuf/repository-service-tuf/issues/41

Closes #281

### CODE REVIEW

**Comment by kairoaraujo** [comment]
@KAUTH its also related to https://github.com/repository-service-tuf/repository-service-tuf/pull/364

File: tests/bdd/tokens/test_generate.py
Comment (MVrachev) [inline tests/bdd/tokens/test_generate.py:36]: I cannot find where is the `token` fixture nor do I see it as a variable in the BDD scenario.
Where is implemented?

File: tests/bdd/tokens/test_generate.py
Comment (MVrachev) [inline tests/bdd/tokens/test_generate.py:33]: I think this should be changed to:
``` suggestion
    parse("the admin adds Authorization Bearer 'access_token' in the 'headers'"),
```

File: tests/bdd/tokens/test_generate.py
Comment (MVrachev) [inline tests/bdd/tokens/test_generate.py:98]: Those will be calculated with the new values of the new `Examples` table, correct?

File: tests/bdd/features/tokens/generate.feature
Comment (KAUTH) [reply tests/bdd/features/tokens/generate.feature:47]: This is interesting, the tests do pass with the error code of `422` because the scopes given in the `Examples` fail the validation from the [`TokenRequestPayload`](https://github.com/repository-service-tuf/repository-service-tuf-api/blob/5204a7095f32672a927b41e4e5e975abd42eff38/repository_service_tuf_api/api/token.py#L58) Pydantic model.  In this case, we allow the admin to create new tokens (for distribution to others), but that can only have specific [limited scope](https://github.com/repository-service-tuf/repository-service-tuf-api/blob/3b40a7e44696f471a72ce7ef5b2dde2f0f2ea58c/repository_service_tuf_api/token.py#L88), and not the ability for example to `write:token` (which only admin should have). Thus, I think it makes sense that it's a validation error, since even with admin permissions you cannot create a token with these scopes.
In contrast, the `/api/v1/token` uses a `application/x-www-form-urlencoded` content type, and validation actually happens in [`token.post`](https://github.com/repository-service-tuf/repository-service-tuf-api/blob/3b40a7e44696f471a72ce7ef5b2dde2f0f2ea58c/repository_service_tuf_api/token.py#L170). For this case, you should not be able to create a token with scope that is not in the user.scopes, so it makes sense to be forbidden I think.

**Review by KAUTH — CHANGES_REQUESTED** [review]

File: tests/bdd/features/tokens/generate.feature
Comment (MVrachev) [reply tests/bdd/features/tokens/generate.feature:47]: Ahaaa, I see. That explains it.
It's not always visible where `pydanitc` validates the things, now when I look at it it's clear.

File: tests/bdd/tokens/test_generate.py
Comment (MVrachev) [reply tests/bdd/tokens/test_generate.py:36]: Hm, I thought the description should exactly match the one described in the text, but it seems that you can assign a value to it that way.

**Comment by kairoaraujo** [comment]
About this PR, as I discussed with @KAUTH at EuroPython and also with @MVrachev before

Should we merge it as an "as is" from the FT in the umbrella and work on specific issues for improvement?
Unless there is a must to fix the comments above.

**Comment by kairoaraujo** [comment]
Rebased

**Comment by KAUTH** [comment]
> Should we merge it as an "as is" from the FT in the umbrella and work on specific issues for improvement?
Unless there is a must to fix the comments above.

Sounds good to me, we can create an issue to work on the comments left in this PR. Also need the tests to succeed before merging.

**Comment by kairoaraujo** [comment]
> Sounds good to me, we can create an issue to work on the comments left in this PR. Also need the tests to succeed before merging.

Fixed the dependencies and tests run.

**Comment by kairoaraujo** [comment]
Rebased

File: tests/bdd/features/tokens/generate.feature
Comment (MVrachev) [reply tests/bdd/features/tokens/generate.feature:69]: This scenario was removed.

File: tests/bdd/features/tokens/login.feature
Comment (MVrachev) [reply tests/bdd/features/tokens/login.feature:52]: This scenario was removed.

File: tests/bdd/tokens/test_generate.py
Comment (MVrachev) [reply tests/bdd/tokens/test_generate.py:33]: This works for now because of the reason described in @KAUTH comment here: https://github.com/repository-service-tuf/repository-service-tuf-api/pull/369#discussion_r1242615306

File: tests/bdd/tokens/test_generate.py
Comment (MVrachev) [reply tests/bdd/tokens/test_generate.py:98]: @kairoaraujo am I correct here?

**Review by MVrachev — COMMENTED** [review]
I created an issue where I tried to link to the improvements comments:
https://github.com/repository-service-tuf/repository-service-tuf-api/issues/398
@KAUTH please have a look if they represent your suggestions accurately or am I missing something.

I only have two questions.
When they are answered I think we are good to go.

**Comment by KAUTH** [comment]
> @KAUTH please have a look if they represent your suggestions accurately or if am I missing something.

@MVrachev I double-checked, you got them all, thanks!

File: tests/bdd/features/tokens/login.feature
Comment (KAUTH) [reply tests/bdd/features/tokens/login.feature:1]: > I think we don't check that in our code as well.

I believe there's validation done in the model [here](https://github.com/repository-service-tuf/repository-service-tuf-api/blob/3b40a7e44696f471a72ce7ef5b2dde2f0f2ea58c/repository_service_tuf_api/token.py#L33-L37), which is defined for the endpoint [here](https://github.com/repository-service-tuf/repository-service-tuf-api/blob/5204a7095f32672a927b41e4e5e975abd42eff38/repository_service_tuf_api/api/token.py#L42).

**Comment by kairoaraujo** [comment]
Rebased

**Review by MVrachev — APPROVED** [review]

**Comment by MVrachev** [comment]
@KAUTH I cannot merge until you give approval as well.
It's showing that 1 person requires changes.

**Review by KAUTH — APPROVED** [review]

**Pull request merged**
